[LTC] Refactor namespace bridge

alanwaketan opened the pull request:
Summary:
This pull request refactors namespace bridge:
1. Removes unused functions from namespace bridge.
2. Merge and prune some bridge utils.
3. Move at::Tensor <=> LazyTensor utilities to tensor.h
3. WIP: holding on #67869 to move device related utils out of bridge.

Test Plan:
lazy_tensor_core/test/cpp/build/test_ptltc

P.S. It's easier to review one commit by another.

Commit:
[LTC] Removes unused functions from namespace bridge

Summary:
This commit removes unused functions from namespace bridge as the first
step to refactor bridge.

Test Plan:
lazy_tensor_core/test/cpp/build/test_ptltc

Commit:
[LTC] Merge and prune some bridge utils

Summary:
This commit merges and prunes some bridge utils:
1. Modifies TryGetLtcTensor to return LazyTensor directly given it's a pointer type anyway.
2. Moves some utils out given they are only useful for a particular file.
3. Adds operator bool() to LazyTensor.

Test Plan:
lazy_tensor_core/test/cpp/build/test_ptltc

Commit:
[LTC] Move at::Tensor <=> LazyTensor utilities to tensor.h

Summary:
This commit moves at::Tensor <=> LazyTensor utilities from bridge to tensor.h.

Test Plan:
lazy_tensor_core/test/cpp/build/test_ptltc

JackCaoG commented:
What's the motivation of moving `aten->LazyTensor` part out of the bridge? I felt like the point of `bridge` is to handle all of the `pytorch core -> Lazy` conversion, inclduing `Tensor` and `Device`. This way in Tensor.h/cpp user should not care about most of the `aten` level stuff.
 
The line is a bit blur right now as `c10:device` is being used more broadly in lazy core but for `at::Tensor -> LazyTensor` above statement should still remain true..

alanwaketan commented:
> What's the motivation of moving `aten->LazyTensor` part out of the bridge? I felt like the point of `bridge` is to handle all of the `pytorch core -> Lazy` conversion, inclduing `Tensor` and `Device`. This way in Tensor.h/cpp user should not care about most of the `aten` level stuff.
> 
> The line is a bit blur right now as `c10:device` is being used more broadly in lazy core but for `at::Tensor -> LazyTensor` above statement should still remain true..

As you pointed out that after the refactor of BackendDevice, most of those device related utils won't make any sense anymore. Then it's logical to move the at::Tensor <=> LazyTensor utils to Tensor.h and delete the bridge concept. You probably haven't notice that LTC's Tensor.h is more concise.

wconstab commented:
@JackCaoG are you ok with this? 

I think a while back our team discussed the tradeoffs of (a) moving bridge methods into LazyTensor as static mehods, (b) moving static methods for create/get tensor out of LazyTensor into bridge.  We basically couldn't find a reason for one way or the other at that time, but my overall sentiment was that it made sense to put all the tensor create/convert methods into one place.

Regarding 
> Tensor.h/cpp user should not care about most of the aten level stuff

This is a good point to discuss; from my perspective, part of our goal was to clean up any types we can remove and use pytorch/aten types instead to simplify the codebase.  To this extent, I didn't see a point in isolating the aten types in the bridge, but just prefer to use them throughout the core.  Of course inside backend code, other types could prevail.  But do you see this as an issue?

JackCaoG reviewed (approved):
Well, I guess having the `at::Tensor -> LazyTensor` in the `tensor.h` also makes sense, especially given the device bridge will be moved to the backend.

Commit:
[LTC] Rename AtenFromLtcTensor to CreateAtenFromLtcTensor

Summary:
This commit renames AtenFromLtcTensor to CreateAtenFromLtcTensor.

Test Plan:
lazy_tensor_core/test/cpp/build/test_ptltc

alanwaketan commented:
Thanks Jack and Will for approving this pull request.
